addons/im_livechat/static/src/embed/common/livechat_service.js

```diff
@@ -131,23 +131,34 @@ export class LivechatService {
         }
         const temporaryThread = this.thread;
         await this._createThread({ persist: true });
-        if (temporaryThread) {
+        const deleteTemporary = async () => {
             await this.store.chatHub.initPromise;
-            const chatWindow = this.store.ChatWindow.get({ thread: temporaryThread });
-            await chatWindow.close({ force: true });
-            temporaryThread.delete();
-        }
+            await this.store.ChatWindow.get({ thread: temporaryThread })?.close({ force: true });
+            temporaryThread?.delete();
+        };
         if (!this.thread) {
+            await deleteTemporary();
             return;
         }
         this.busService.addChannel(`mail.guest_${this.guestToken}`);
-        await Promise.all([
-            this.thread.openChatWindow({ focus: true }),
-            this.env.services["mail.store"].initialize(),
-        ]);
+        await this.thread.fetchNewMessages();
+        this.thread.firstActionAfterPersistDeferred.then(async () => {
+            await Promise.all([
+                this.store.chatHub.initPromise,
+                this.env.services["mail.store"].initialize(),
+                this.thread.isLoadedDeferred,
+            ]);
+            // Do not load unread messaes: new messages were loaded to avoid
+            // flickering, we do not want another load that would cause the same
+            // issue.
+            this.thread.scrollUnread = false;
+            await Promise.all([deleteTemporary(), this.thread.openChatWindow({ focus: true })]);
```

Comment:
I've not tested if it makes any difference but shouldn't we change the thread of the existing chat window before deleting the temporary one, rather than opening a new one?

Reply:
Swap is invisible, but it would be better if we were able to change the thread instead of creating/delete chat windows. I'll try it out.